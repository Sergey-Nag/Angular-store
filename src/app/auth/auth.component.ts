import { Component, OnDestroy, OnInit } from '@angular/core';
import { AuthService } from './auth.service';

@Component({
  selector: 'app-auth',
  templateUrl: './auth.component.html',
  styleUrls: ['./auth.component.scss']
})
export class AuthComponent implements OnDestroy, OnInit {

  constructor(private authService: AuthService) {
  }

  ngOnInit() {
    this.authService.setAuth(false);
  }

  ngOnDestroy() {
    this.authService.setAuth(true);
  }

}
