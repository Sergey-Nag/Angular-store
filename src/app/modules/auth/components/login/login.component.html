<div class="container">
    <div class="row h-100 justify-content-center align-items-center">
        <div class="col-sm-6 col-xl-4">
            <div class="card">
                <div class="card-body">
                    <img src="../../../assets/unknown-guy.svg" alt="avatar" class="w-100">
                    <hr>
                    <form [formGroup]="authForm" (ngSubmit)="handleSubmit()">
                        <label for="username" class="form-label">Username</label>
                        <input
                            type="text"
                            id="username"
                            class="form-control"
                            name="username"
                            formControlName="username"
                            [class.is-invalid]="isControlInvalid('username')"
                            [class.is-valid]="authForm.dirty && !isControlInvalid('username')">
                        <div class="invalid-feedback">
                            <p *ngFor="let error of getControlErrorMessage('username')" class="m-0">
                                {{error}}
                            </p>
                        </div>
                        <div class="d-grid mt-3">
                            <button
                                type="submit"
                                class="btn btn-primary"
                                [disabled]="authForm.invalid || isLoading">
                                    <span *ngIf="isLoading" class="spinner-border spinner-border-sm" role="status" aria-hidden="true"></span>
                                    Log in
                                </button>
                        </div>
                    </form>
                </div>
            </div>
        </div>
    </div>
</div>