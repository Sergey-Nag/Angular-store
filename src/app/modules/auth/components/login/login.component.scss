.container {
    height: 100vh;
}

.form-control.ng-touched {
    &.ng-invalid {
        border-color: var(--bs-danger);
    }
    &.ng-valid {
        border-color: var(--bs-success);
    }
}