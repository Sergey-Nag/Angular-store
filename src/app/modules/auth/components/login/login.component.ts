import { Component } from '@angular/core';
import { FormControl, FormGroup, NgForm, Validators } from '@angular/forms';
import { Router } from '@angular/router';
import { VALIDATION_ERROR } from '@shared/helpers/validation-error.helper';
import { debounceTime } from 'rxjs/operators';
import { AuthService } from '../../services/auth.service';

@Component({
  selector: 'app-login',
  templateUrl: './login.component.html',
  styleUrls: ['./login.component.scss']
})
export class LoginComponent {
  isLoading = false;
  authForm = new FormGroup({
    username: new FormControl('', [
      Validators.required,
      Validators.minLength(4),
      Validators.email,
      Validators.maxLength(40),
    ])
  }, { updateOn: 'blur'});

  constructor(private authService: AuthService, private router: Router) {}

  isControlInvalid(controlName: string): boolean {
    return this.authForm.controls[controlName].touched && this.authForm.controls[controlName].invalid;
  }

  getControlErrorMessage(controlName: string): string[] {
    return Object.keys(
        this.authForm.controls[controlName].errors || []
      )
      .map((errorKey) => VALIDATION_ERROR[errorKey]);
  }

  handleSubmit() {
    if (this.authForm.invalid) return;
    
    this.startLoading();

    this.authService
      .signIn(this.authForm.value.username)
      .subscribe(() => {
        this.finishLoading();
        this.router.navigate(['/']);
      });
  }

  private startLoading() {
    this.isLoading = true;
    this.authForm.get('username').disable();
  }

  private finishLoading() {
    this.isLoading = false;
    this.authForm.get('username').enable();
  }
}
