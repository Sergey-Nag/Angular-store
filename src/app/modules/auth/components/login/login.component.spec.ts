import { ReactiveFormsModule } from "@angular/forms";
import { Router } from "@angular/router";
import { VALIDATION_ERROR } from "@shared/helpers/validation-error.helper";
import { fireEvent, getByText, render, screen, RenderResult } from "@testing-library/angular";
import { waitFor } from "@testing-library/dom";
import { of } from "rxjs";
import { AuthService } from "../../services/auth.service";
import { LoginComponent } from "./login.component";

const renderWithServices = async () => 
  await render(LoginComponent, {
    imports: [ReactiveFormsModule],
    providers: [
      {
        provide: AuthService,
        useValue: {
          signIn: jest.fn().mockReturnValue(of({}))
        }
      },
      {
        provide: Router,
        useValue: {
          navigate: jest.fn()
        }
      }
    ]
  });

describe('Login form', () => {
  it('Should render login form component', async () => {
    const { fixture } = await renderWithServices();

    expect(fixture).toMatchSnapshot();
  });

  it('Should be invalid and show "Required" error message with the empty input value', async () => {
    const { fixture } = await renderWithServices();
    const input = screen.getByRole('textbox');

    fireEvent.input(input, { target: { value: '' }});
    fireEvent.blur(input);

    const errorMessage = screen.getByText(VALIDATION_ERROR.required);

    expect(fixture.componentInstance.authForm.invalid).toBe(true);

    expect(errorMessage).toBeInTheDocument();
    expect(errorMessage).toBeVisible();
    expect(screen.getByRole('button', { name: 'Log in'})).toBeDisabled();
  });

  it('Should be invalid and show "Min Length" error message with the "abc" input value', async () => {
    const { fixture } = await renderWithServices();
    const input = screen.getByRole('textbox');

    fireEvent.input(input, { target: { value: 'abc' }});
    fireEvent.blur(input);

    const errorMessage = await waitFor(() => screen.getByText(VALIDATION_ERROR.minlength));

    expect(fixture.componentInstance.authForm.invalid).toBe(true);

    expect(errorMessage).toBeInTheDocument();
    expect(errorMessage).toBeVisible();
    expect(screen.getByRole('button', { name: 'Log in'})).toBeDisabled();
  });

  it('Should be invalid and show "Max Length" error message with the input value more than 40 symbols', async () => {
    const { fixture } = await renderWithServices();
    const input = screen.getByRole('textbox');

    fireEvent.input(input, { target: { value: 'abcdefgabcdefgabcdefgabcdefgabcdefgabcdefg' }});
    fireEvent.blur(input);

    const errorMessage = await waitFor(() => screen.getByText(VALIDATION_ERROR.maxlength));

    expect(fixture.componentInstance.authForm.invalid).toBe(true);

    expect(errorMessage).toBeInTheDocument();
    expect(errorMessage).toBeVisible();
    expect(screen.getByRole('button', { name: 'Log in'})).toBeDisabled();
  });
  
  describe('Should be invalid and show "Email" error message with the different invalid emails', () => {
    it('The "12345" value should be invalid', async () => {
      const { fixture } = await renderWithServices();
      const input = screen.getByRole('textbox');

      fireEvent.input(input, { target: { value: '12345' }});
      fireEvent.blur(input);
  
      const errorMessage = await waitFor(() => screen.getByText(VALIDATION_ERROR.email));
  
      expect(fixture.componentInstance.authForm.invalid).toBe(true);
  
      expect(errorMessage).toBeVisible();
      expect(screen.getByRole('button', { name: 'Log in'})).toBeDisabled();
    });
  
    it('The "abc@" value should be invalid', async () => {
      const { fixture } = await renderWithServices();
      const input = screen.getByRole('textbox');

      fireEvent.input(input, { target: { value: 'abc@' }});
      fireEvent.blur(input);
  
      const errorMessage = await waitFor(() => screen.getByText(VALIDATION_ERROR.email));
  
      expect(fixture.componentInstance.authForm.invalid).toBe(true);
  
      expect(errorMessage).toBeVisible();
      expect(screen.getByRole('button', { name: 'Log in'})).toBeDisabled();
    });

    it('The "@abc.de" value should be invalid', async () => {
      const { fixture } = await renderWithServices();
      const input = screen.getByRole('textbox');

      fireEvent.input(input, { target: { value: '@abc.de' }});
      fireEvent.blur(input);
  
      const errorMessage = await waitFor(() => screen.getByText(VALIDATION_ERROR.email));
  
      expect(fixture.componentInstance.authForm.invalid).toBe(true);
  
      expect(errorMessage).toBeVisible();
      expect(screen.getByRole('button', { name: 'Log in'})).toBeDisabled();
    });
  });

  it('Should be valid with the correct email and don\'t show error message', async () => {
    const { fixture, getByRole } = await renderWithServices();
    const input = getByRole('textbox');

    fireEvent.input(input, { target: { value: 'test@test.t' }});
    fireEvent.blur(input);

    expect(fixture.componentInstance.authForm.valid).toBe(true);
    expect(screen.getByRole('button', { name: 'Log in'})).toBeEnabled();
  });

  it('Should call signIn method from AuthService and redirect to the main route after submit for with valid data', async () => {
    const { fixture } = await renderWithServices();
    const authService = fixture.componentRef.injector.get(AuthService);
    const router = fixture.componentRef.injector.get(Router);

    const input = screen.getByRole('textbox');
    const signInButton = screen.getByRole('button', { name: 'Log in'});

    fireEvent.input(input, { target: { value: 'test@test.t' }});
    fireEvent.blur(input);

    fireEvent.click(signInButton);

    expect(authService.signIn).toBeCalled();
    expect(router.navigate).toBeCalledWith(['/']);
  });
});