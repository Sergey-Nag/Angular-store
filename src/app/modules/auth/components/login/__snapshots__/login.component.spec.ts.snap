// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login form Should render login form component 1`] = `
<app-login
  __ngContext__={[Function LRootView]}
  authForm={[Function FormGroup]}
  authService={[Function Object]}
  isLoading="false"
  router={[Function Object]}
>
  <div
    class="container"
  >
    <div
      class="row h-100 justify-content-center align-items-center"
    >
      <div
        class="col-sm-6 col-xl-4"
      >
        <div
          class="card"
        >
          <div
            class="card-body"
          >
            <img
              alt="avatar"
              class="w-100"
              src="../../../assets/unknown-guy.svg"
            />
            <hr />
            <form
              class="ng-untouched ng-pristine ng-invalid"
              novalidate=""
            >
              <label
                class="form-label"
                for="username"
              >
                Username
              </label>
              <input
                class="form-control ng-untouched ng-pristine ng-invalid"
                formcontrolname="username"
                id="username"
                name="username"
                type="text"
              />
              <div
                class="invalid-feedback"
              >
                <p
                  class="m-0"
                >
                   Field is required 
                </p>
              </div>
              <div
                class="d-grid mt-3"
              >
                <button
                  class="btn btn-primary"
                  disabled=""
                  type="submit"
                >
                   Log in 
                </button>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</app-login>
`;
